{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CoV Pan-genome test alignments\n",
    "```\n",
    "Lead     : Ababaian\n",
    "Issue    : #27\n",
    "start    : 2020 04 07\n",
    "complete : 2020 04 07\n",
    "files    : s3://serratus-public/notebook/200407/\n",
    "```\n",
    "\n",
    "## Introduction\n",
    "We have an initial version of the CoV pan-genome, `cov0r`. As an initial sanity check, align a known not-CoV transcriptome, a genome, and a CoV+ transcriptome against `cov0r` to see how well the alignments 'stick'. More importantly how much bleed is there from a mammalian transcriptome into the `cov0r` pan-genome.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Materials and Methods\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequence Accessions/Sources\n",
    "\n",
    "- CoV Index: `s3://serratus-public/seq/cov0/cov0r.fa`\n",
    "- Human WGS reads: `s3://serratus-public/test-data/fq/NA12878.1.fq.gz`\n",
    "- Cat transcriptome: `fasterq-dump SRR6639048`\n",
    "- SARS-CoV-2 patient meta-transcriptome: `fasterq-dump SRR11454613`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## C4.large Instance\n",
    "## in serratus-align v0.2 container\n",
    "\n",
    "# Magic Blast alignments\n",
    "GENOME='cov0r'\n",
    "ALG_ARG='-splice F -no_unaligned -max_db_word_count 1000000'\n",
    "\n",
    "## Human WGS\n",
    "magicblast -infmt fastq $ALG_ARG  \\\n",
    "  -query NA12878.1.fq.gz \\\n",
    "  -db $GENOME |\\\n",
    "  samtools view -b - > NA12878.mb.bam\n",
    "  \n",
    "## Cat RNAseq\n",
    "magicblast -infmt fastq $ALG_ARG  \\\n",
    "  -query SRR6639048.fastq \\\n",
    "  -db $GENOME |\\\n",
    "  samtools view -b - > SRR6639048.mb.bam\n",
    "\n",
    "## CoV+\n",
    "magicblast -infmt fastq $ALG_ARG  -paired \\\n",
    "  -query SRR11454613_1.fastq -query_mate SRR11454613_2.fastq  \\\n",
    "  -db $GENOME | \\\n",
    "  samtools view -b - > SRR11454613.mb.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## C4.large Instance\n",
    "## in serratus-align v0.2 container\n",
    "\n",
    "# Bowtie2 alignments\n",
    "GENOME='cov0r'\n",
    "ALG_ARG='--very-sensitive-local'\n",
    "\n",
    "## Human WGS\n",
    "bowtie2 $BT2_ARG \\\n",
    "  -x $GENOME \\\n",
    "  -U NA12878.1.fq.gz | \\\n",
    "  samtools view -bS - > NA12878.bt2.bam\n",
    "  \n",
    "## Cat RNAseq\n",
    "bowtie2 $BT2_ARG \\\n",
    "  -x $GENOME \\\n",
    "  -U SRR6639048.fastq | \\\n",
    "  samtools view -bS - > SRR6639048.bt2.bam\n",
    "\n",
    "## CoV+\n",
    "bowtie2 $BT2_ARG \\\n",
    "  -x $GENOME \\\n",
    "  -1 SRR11454613_1.fastq -2 SRR11454613_2.fastq | \\\n",
    "  samtools view -bS - > SRR11454613.bt2.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Flagstat output from alignments\n",
    "\n",
    "### Human WGS\n",
    "\n",
    "```\n",
    "NA12878.bt2.bam\n",
    "840909 + 0 in total (QC-passed reads + QC-failed reads)\n",
    "0 + 0 secondary\n",
    "0 + 0 supplementary\n",
    "0 + 0 duplicates\n",
    "206 + 0 mapped (0.02% : N/A)\n",
    "0 + 0 paired in sequencing\n",
    "0 + 0 read1\n",
    "0 + 0 read2\n",
    "0 + 0 properly paired (N/A : N/A)\n",
    "0 + 0 with itself and mate mapped\n",
    "0 + 0 singletons (N/A : N/A)\n",
    "0 + 0 with mate mapped to a different chr\n",
    "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
    "\n",
    "NA12878.mb.bam\n",
    "0 + 0 in total (QC-passed reads + QC-failed reads)\n",
    "0 + 0 secondary\n",
    "0 + 0 supplementary\n",
    "0 + 0 duplicates\n",
    "0 + 0 mapped (N/A : N/A)\n",
    "0 + 0 paired in sequencing\n",
    "0 + 0 read1\n",
    "0 + 0 read2\n",
    "0 + 0 properly paired (N/A : N/A)\n",
    "0 + 0 with itself and mate mapped\n",
    "0 + 0 singletons (N/A : N/A)\n",
    "0 + 0 with mate mapped to a different chr\n",
    "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
    "```\n",
    "\n",
    "### Cat RNAseq\n",
    "```\n",
    "SRR6639048.bt2.bam\n",
    "31886603 + 0 in total (QC-passed reads + QC-failed reads)\n",
    "0 + 0 secondary\n",
    "0 + 0 supplementary\n",
    "0 + 0 duplicates\n",
    "169 + 0 mapped (0.00% : N/A)\n",
    "0 + 0 paired in sequencing\n",
    "0 + 0 read1\n",
    "0 + 0 read2\n",
    "0 + 0 properly paired (N/A : N/A)\n",
    "0 + 0 with itself and mate mapped\n",
    "0 + 0 singletons (N/A : N/A)\n",
    "0 + 0 with mate mapped to a different chr\n",
    "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
    "\n",
    "SRR6639048.mb.bam\n",
    "217 + 0 in total (QC-passed reads + QC-failed reads)\n",
    "47 + 0 secondary\n",
    "0 + 0 supplementary\n",
    "0 + 0 duplicates\n",
    "217 + 0 mapped (100.00% : N/A)\n",
    "0 + 0 paired in sequencing\n",
    "0 + 0 read1\n",
    "0 + 0 read2\n",
    "0 + 0 properly paired (N/A : N/A)\n",
    "0 + 0 with itself and mate mapped\n",
    "0 + 0 singletons (N/A : N/A)\n",
    "0 + 0 with mate mapped to a different chr\n",
    "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
    "```\n",
    "\n",
    "### CoV metaRNAseq\n",
    "```\n",
    "SRR11454613.bt2.bam\n",
    "10837378 + 0 in total (QC-passed reads + QC-failed reads)\n",
    "0 + 0 secondary\n",
    "0 + 0 supplementary\n",
    "0 + 0 duplicates\n",
    "451808 + 0 mapped (4.17% : N/A)\n",
    "10837378 + 0 paired in sequencing\n",
    "5418689 + 0 read1\n",
    "5418689 + 0 read2\n",
    "263554 + 0 properly paired (2.43% : N/A)\n",
    "404092 + 0 with itself and mate mapped\n",
    "47716 + 0 singletons (0.44% : N/A)\n",
    "131098 + 0 with mate mapped to a different chr\n",
    "196 + 0 with mate mapped to a different chr (mapQ>=5)\n",
    "\n",
    "SRR11454613.mb.bam\n",
    "61615 + 0 in total (QC-passed reads + QC-failed reads)\n",
    "42263 + 0 secondary\n",
    "0 + 0 supplementary\n",
    "0 + 0 duplicates\n",
    "61615 + 0 mapped (100.00% : N/A)\n",
    "19352 + 0 paired in sequencing\n",
    "10061 + 0 read1\n",
    "9291 + 0 read2\n",
    "8266 + 0 properly paired (42.71% : N/A)\n",
    "8296 + 0 with itself and mate mapped\n",
    "11056 + 0 singletons (57.13% : N/A)\n",
    "70 + 0 with mate mapped to a different chr\n",
    "70 + 0 with mate mapped to a different chr (mapQ>=5)\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert to only mapped reads and their pairs\n",
    "# for each library\n",
    "samtools view -b -F 4 NA12878.bt2.bam > NA12878.bt.bam\n",
    "mv NA12878.bt.bam NA12878.bt2.bam\n",
    "\n",
    "samtools view -b -F 4 SRR6639048.bt2.bam > SRR6639048.bt.bam\n",
    "mv SRR6639048.bt.bam SRR6639048.bt2.bam\n",
    "\n",
    "samtools view -b -G 12 SRR11454613.bt2.bam > SRR11454613.bt.bam\n",
    "mv SRR11454613.bt.bam SRR11454613.bt2.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In dir with only bam files\n",
    "aws s3 cp --recursive ./ s3://serratus-public/notebook/20200407/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results & Discussion\n",
    "\n",
    "Sources of error in analysis\n",
    "\n",
    "### KC786228.1\n",
    "\n",
    "- BT2 gave 206 aligned reads from human WGS to `cov0` that MB did not. All but one read were mapped to `KC786228.1`. In the human genome this maps to ribosomal DNA. Most likely some kind of rDNA contaminant in accession `KC786228.1: UNVERIFIED: Infectious bronchitis virus isolate YN/V90/2012 nucleocapsid protein-like (N) gene, partial sequence`. Recommend removal of this accession.\n",
    "\n",
    "```\n",
    "ERR194147.505740963     0       KC786228.1      94      0       101M    *       0       0       ACGGGTTACCCGCGCCTGCCGGCGTAGGGTAGGCACACGCTGAGCCAGTCAGTGTAGCGCGCGTGCAGCCCCGGACATCTAAGGGCATCACAGACCTGTTA   CCCFFDFFHHHHHIIJJJJJIIJIGIIJJ@FEHIJIHHGFFDDDEDDDCDDDCC@DDDDDDDDBDDDDDDDDDDB<@CCCDDDDDDDDDDDDDDDDDD@C@   AS:i:-57        XN:i:0  XM:i:11 XO:i:0  XG:i:0  NM:i:11 MD:Z:0G23G1T5A1T2A18A0T27T6C5C2 YT:Z:UU\n",
    "ERR194147.505741305     0       KC786228.1      94      0       101M    *       0       0       ACGGGTTACCCGCGCCTGCCGGCGTAGGGTAGGCACACGCTGAGCCAGTCAGTGTAGCGCGCGTGCAGCCCCGGACATCTAAGGGCATCACAGACCTGTTA   CCCFFFFFHHHHHJJJJJJJJJIJHIJJJFHIIEHGHHHFFDDDDDDDDDDDDDDFDDDDDDDBDDDDCABDDDB9>CCDDDDDDDDDCCDDDDDDDDCDC   AS:i:-57        XN:i:0  XM:i:11 XO:i:0  XG:i:0  NM:i:11 MD:Z:0G23G1T5A1T2A18A0T27T6C5C2 YT:Z:UU\n",
    "ERR194147.505741455     0       KC786228.1      97      0       101M    *       0       0       GGTTACCCGCGCCTGCCGGCGTAGGGTAGGCACACGCTGAGCCAGTCAGTGTAGCGCGCGTGCAGCCCCGGACATCTAAGGGCATCACAGACCTGTTATTG   @@CFFFFFHDHGHJJJIJIFI@:FHHBBCGGGGHGGHEFDEFEEE>CCCCC>B>>==B;B<77A@@BBDBD@B>C@CCDCABBDCCDDDDC?BC<+4::@:   AS:i:-59        XN:i:0  XM:i:13 XO:i:0  XG:i:0  NM:i:13 MD:Z:21G1T5A1T2A18A0T27T6C5C2C0A0A0     YT:Z:UU\n",
    "ERR194147.612172157     0       KC786228.1      93      0       101M    *       0       0       AACGGGTTACCCGCGCCTGCCGGCGTAGGGTAGGCACACGCTGAGCCAGTCAGTGTAGCGCGCGTGCAGCCCCGGACATCTAAGGGCATCACAGACCTGTT   CCCFFFFFHHHHGJIJJJJJJIIIIBGGJJ@EEIFHHEEEFACDDCDDDCCCDCDDFDDDDDDD@DDDDDDDDDDD>@CCDDDCDDDDDDDD>CCBDCCCD   AS:i:-61        XN:i:0  XM:i:12 XO:i:0  XG:i:0  NM:i:12 MD:Z:0C0G23G1T5A1T2A18A0T27T6C5C1       YT:Z:UU\n",
    "```\n",
    "\n",
    "### AX191447.1 and AX191449.1\n",
    "\n",
    "These are labelled as from Rat Coronavirus but both correspond to a patent filing related to rat  Xylosyltransferase. Exclude from cov0. Can confirm that these reads when blat against human map to `YXLT2`.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
